llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2022 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+import logging
+import os
+from typing import Any, Dict, List, Optional, Union
+
+from composer.core import Callback, Event, State, Time
+from composer.loggers import Logger
+from composer.loggers.mosaicml_logger import (MOSAICML_PLATFORM_ENV_VAR,
+                                              RUN_NAME_ENV_VAR)
+from composer.utils import dist
+from composer.utils.misc import create_interval_scheduler
+
+from mcli import ComputeConfig, Run, RunConfig, create_run, get_run
+
+log = logging.getLogger(__name__)
+
+MAX_RUN_NAME_LENGTH = 40
+
+REQUIRED_PARAMS_FOR_EVAL = {
+    'device_eval_batch_size',
+    'icl_tasks',  # only required for eval, may not be specified in pure training
+    'max_seq_len',
+    'model',  # converted into models
+    'tokenizer',  # converted into models
+    'save_folder',  # required, but used as load_path
+}
+OPTIONAL_PARAMS_FOR_EVAL = {
+    'dist_timeout',
+    'eval_gauntlet',
+    'fsdp_config',
+    'icl_subset_num_batches',
+    'loggers',
+    'precision',
+    'python_log_level',
+    'seed',
+}
+
+
+def get_run_name(previous_run_name: str, count: int) -> str:
+    *name_without_uuid_suffix, _ = previous_run_name.split('-')
+    name_suffix = ('-'.join(name_without_uuid_suffix))
+
+    if len(name_suffix) > MAX_RUN_NAME_LENGTH:
+        log.warning(
+            f'Training run name {name_suffix} may be too long, truncating to {MAX_RUN_NAME_LENGTH} characters'
+        )
+        name_suffix = name_suffix[:MAX_RUN_NAME_LENGTH]
+
+    return f'eval{count}-{name_suffix}'
+
+
+def get_load_path(save_folder: str,
+                  save_latest_filename: Optional[str] = None) -> str:
+    # TODO: check that the prefix is remote and not a local file (not supported of course)
+
+    if not save_latest_filename:
+        rank = dist.get_global_rank()
+        save_latest_filename = f'latest-rank{rank}.pt'
+
+    return f'{save_folder}/{save_latest_filename}'
+
+
+def get_eval_models_dict(
+    model: Dict[str, Any],
+    tokenizer: Dict[str, Any],
+) -> List[Dict[str, Any]]:
+    name = model.get('name')
+
+    cfg_overrides = model.pop('cfg_overrides', {})
```

Comment:
this doesn't seem quite right, could you explain what you are trying to do with these overrides?

Reply:
This was a slight typo on the key name (->`config_overrides`). I wasn't seeing `config_overrides` being used by the eval yamls?

For example, random training yaml
```yaml
model:
  name: hf_causal_lm
  pretrained_model_name_or_path: mosaicml/mpt-7b
  pretrained: true  # false: only use the architecture; true: initialize with pretrained weights
  config_overrides:
    max_seq_len: ${max_seq_len}
    attn_config:
      attn_impl: triton
      # Set this to `true` if using `train_loader.dataset.packing_ratio` below
      attn_uses_sequence_id: false
```

model configs in eval yamls:
```yaml
  models:
  -
    model_name: mpt1b
    tokenizer:
      name: EleutherAI/gpt-neox-20b
      kwargs:
        model_max_length: ${max_seq_len}
    model:
      name: mpt_causal_lm
      init_device: mixed
      d_model: 2048
      n_heads: 16 # Modified 24->16 so that d_head == 128 to satisfy FlashAttention
      n_layers: 24
      expansion_ratio: 4
      max_seq_len: ${max_seq_len}
      vocab_size: 50368
      attn_config:
        attn_impl: triton
```

Reply:
it will depend on which model type you are instantiating. `hf_causal_lm` and `mpt_causal_lm` yaml sections look different, but each type should persist from the original training config straight to the eval config. Might be easier to explain more in person.